vncviewer/Viewport.cxx

```diff
@@ -1118,6 +1131,55 @@ int Viewport::handleSystemEvent(void *event, void *data)
         self->handleKeyRelease(0x36);
     }
 
+    return 1;
+  } else if ((msg->message == WM_CHAR) || (msg->message == WM_DEADCHAR) ||
+             (msg->message == WM_SYSCHAR) || (msg->message == WM_SYSDEADCHAR)) {
+    // Windows will send VK_PACKET key events if it doesn't have a physical key
+    // associated with the event (e.g. from a mobile device virtual keyboard).
+    // After translating a keydown/keyup event pair with VK_PACKET as vKey, a
+    // WM_*CHAR message will be issued containing a UTF-16 code unit. Unicode
+    // characters outside of the range U+0000 - U+ffff (Basic Multilingual
+    // Plane, BMP), are encoded as a pair of UTF-16 code units that will need to
+    // be synthesized into one Unicode code point.
+    uint32_t ucsCode = msg->wParam;
+    if ((ucsCode & 0xfc00) == 0xd800) {
+      // We have received a high surrogate code unit. Remember it and wait for
+      // the low surrogate which should come immediately after.
+      if (self->vkPacketHighSurrogate) {
+        vlog.error(_("Unmatched UTF-16 surrogate pair through VK_PACKET, "
+                     "codes: 0x%04x, 0x%04x"),
+                   self->vkPacketHighSurrogate, ucsCode);
+      }
+      self->vkPacketHighSurrogate = ucsCode;
+      return 1;
+    }
+    uint32_t codePoint = 0;
+    if ((ucsCode & 0xfc00) == 0xdc00) {
+      // We have received a low surrogate code unit. We should have a high
+      // surrogate saved that we can use to calculate the code point.
+      if (!self->vkPacketHighSurrogate) {
+        vlog.error(_("Unmatched UTF-16 surrogate pair through VK_PACKET, "
+                     "code: 0x%04x"),
+                   ucsCode);
+        return 1;
+      }
+      codePoint = (((self->vkPacketHighSurrogate & 0x03ff) << 10) |
+                   (ucsCode & 0x03ff)) + 0x010000;
+    } else {
+      // BMP character. Unicode characters in this range are encoded as a
+      // single UTF-16 code unit.
+      if (self->vkPacketHighSurrogate) {
+        vlog.error(_("Unmatched UTF-16 surrogate pair through VK_PACKET, "
+                     "codes: 0x%04x, U+%04x"),
+                   self->vkPacketHighSurrogate, ucsCode);
+        self->vkPacketHighSurrogate = 0;
+      }
+      codePoint = ucsCode;
+    }
+    uint32_t keySym = ucs2keysym(codePoint);
+    uint32_t keyCode = 0x100 + keySym; // Fake key code
```

Comment:
Keysym is 32-bit, so this is not going to work. Perhaps something similar to the universal keysyms? I.e. `0x01000000 | codePoint`?

Reply:
Good point, will change to what you suggested. Another option is actually a constant, e.g. `keyCode = 0x123`, since we expect only one concurrent `VK_PACKET` key down.